src/perf/lib/ThroughputClient.cpp

```diff
@@ -0,0 +1,419 @@
+        WriteOutput("[conn][%p] Shutdown\n", ConnectionHandle);
+        break;
+    case QUIC_CONNECTION_EVENT_SHUTDOWN_COMPLETE:
+        WriteOutput("[conn][%p] All done\n", ConnectionHandle);
+        ConnectionDataAllocator.Free(ConnData);
+        QuicEventSet(StopEvent);
+        break;
+    case QUIC_CONNECTION_EVENT_RESUMPTION_TICKET_RECEIVED:
+        WriteOutput("[conn][%p] Resumption ticket received (%u bytes):\n",
+            ConnectionHandle,
+            Event->RESUMPTION_TICKET_RECEIVED.ResumptionTicketLength);
+        for (uint32_t i = 0; i < Event->RESUMPTION_TICKET_RECEIVED.ResumptionTicketLength; i++) {
+            WriteOutput("%.2X", (uint8_t)Event->RESUMPTION_TICKET_RECEIVED.ResumptionTicket[i]);
+        }
+        WriteOutput("\n");
+        break;
+    default:
+        break;
+    }
+    return QUIC_STATUS_SUCCESS;
+}
+
+QUIC_STATUS
+ThroughputClient::StreamCallback(
+    _In_ HQUIC StreamHandle,
+    _Inout_ QUIC_STREAM_EVENT* Event,
+    _Inout_ StreamData* StrmData
+    ) {
+    switch (Event->Type) {
+    case QUIC_STREAM_EVENT_SEND_COMPLETE: {
+        SendRequest* Req = (SendRequest*)Event->SEND_COMPLETE.ClientContext;
+        if (!Event->SEND_COMPLETE.Canceled) {
+            uint64_t BytesLeftToSend = Length - StrmData->BytesSent;
+            StrmData->BytesCompleted += Req->QuicBuffer.Length;
+            if (BytesLeftToSend != 0) {
+                Req->SetLength(BytesLeftToSend);
+                StrmData->BytesSent += Req->QuicBuffer.Length;
+
+                QUIC_STATUS Status =
+                    MsQuic->StreamSend(
+                        StrmData->Stream.Handle,
+                        &Req->QuicBuffer,
+                        1,
+                        Req->Flags,
+                        Req);
+
+                if (QUIC_SUCCEEDED(Status)) {
+                    Req = nullptr;
+                }
+            }
+        }
+        if (Req) {
+            SendRequestAllocator.Free(Req);
+        }
+        break;
+    }
+    case QUIC_STREAM_EVENT_PEER_SEND_ABORTED:
+    case QUIC_STREAM_EVENT_PEER_RECEIVE_ABORTED:
+
+        MsQuic->StreamShutdown(
+            StreamHandle,
+            QUIC_STREAM_SHUTDOWN_FLAG_ABORT_SEND | QUIC_STREAM_SHUTDOWN_FLAG_ABORT_RECEIVE,
+            0);
+        break;
+    case QUIC_STREAM_EVENT_SHUTDOWN_COMPLETE: {
+        StrmData->EndTime = QuicTimeUs64();
+        uint64_t ElapsedMicroseconds = StrmData->EndTime - StrmData->StartTime;
+        uint32_t SendRate = (uint32_t)((StrmData->BytesCompleted * 1000 * 1000 * 8) / (1000 * ElapsedMicroseconds));
+
+        WriteOutput("[%p][%llu] Closed [%s] after %u.%u ms. (TX %llu bytes @ %u kbps).\n",
+            StrmData->Connection,
+            GetStreamID(MsQuic, StreamHandle),
+            "Complete",
+            (uint32_t)(ElapsedMicroseconds / 1000),
+            (uint32_t)(ElapsedMicroseconds % 1000),
+            StrmData->BytesCompleted, SendRate);
+
+        StreamDataAllocator.Free(StrmData);
```

Comment:
I assume it is, but make sure this is closing the stream.

Reply:
It is. The destructor will close the stream.